flocker/ca/test/test_ca.py

```diff
@@ -16,13 +16,296 @@
+    """
+    Tests for ``flocker.ca._ca.UserCredential``.
+    """
+    def setUp(self):
+        """
+        Generate a RootCredential for the API certificate tests
+        to work with.
+        """
+        self.path = FilePath(self.mktemp())
+        self.path.makedirs()
+        self.ca = RootCredential.initialize(self.path, b"mycluster")
+        self.username = b"alice"
+
+    def test_written_keypair_exists(self):
+        """
+        ``UserCredential.initialize`` writes a PEM file to the
+        specified path.
+        """
+        uc = UserCredential.initialize(self.path, self.ca, self.username)
+        cert_file = b"{user}.crt".format(user=uc.username)
+        key_file = b"{user}.key".format(user=uc.username)
+        self.assertEqual(
+            (True, True),
+            (self.path.child(cert_file).exists(),
+             self.path.child(key_file).exists())
+        )
+
+    def test_decoded_certificate_matches_public_key(self):
+        """
+        A decoded certificate's public key matches the public key it is
+        meant to be paired with.
+        """
+        uc = UserCredential.initialize(self.path, self.ca, self.username)
+        self.assertTrue(
+            uc.credential.keypair.keypair.matches(
+                uc.credential.certificate.getPublicKey())
+        )
+
+    def test_decoded_certificate_matches_private_key(self):
+        """
+        A decoded certificate matches the private key it is meant to
+        be paired with.
+        """
+        uc = UserCredential.initialize(self.path, self.ca, self.username)
+        priv = uc.credential.keypair.keypair.original
+        pub = uc.credential.certificate.getPublicKey().original
+        pub_asn1 = crypto.dump_privatekey(crypto.FILETYPE_ASN1, pub)
+        priv_asn1 = crypto.dump_privatekey(crypto.FILETYPE_ASN1, priv)
+        pub_der = asn1.DerSequence()
+        pub_der.decode(pub_asn1)
+        priv_der = asn1.DerSequence()
+        priv_der.decode(priv_asn1)
+        pub_modulus = pub_der[1]
+        priv_modulus = priv_der[1]
+        self.assertEqual(pub_modulus, priv_modulus)
+
+    def test_written_keypair_reloads(self):
+        """
+        A keypair written by ``UserCredential.initialize`` can be
+        successfully reloaded in to an identical ``ControlCertificate``
+        instance.
+        """
+        uc1 = UserCredential.initialize(self.path, self.ca, self.username)
+        uc2 = UserCredential.from_path(self.path, uc1.username)
+        self.assertEqual(uc1, uc2)
+
+    def test_keypair_correct_umask(self):
+        """
+        A keypair file written by ``UserCredential.initialize`` has
+        the correct permissions (0600).
+        """
+        uc = UserCredential.initialize(self.path, self.ca, self.username)
+        key_file = b"{user}.key".format(user=uc.username)
+        keyPath = self.path.child(key_file)
+        st = os.stat(keyPath.path)
+        self.assertEqual(b'0600', oct(st.st_mode & 0777))
```

Comment:
Some of these tests aren't for the key/certificate handling logic exactly.  These tests are very slow (>1s each on average - that's at least 100x too slow) because of the large number of large key pairs they generate.  It would be nice to avoid doing all of this extra work just to make an assertion about the permissions of the file the data gets written to.